File: c++/66_Plus_One.cpp
#include <vector>
using namespace std;

// Beat 100% in time and 28% in memory
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {

        int i = digits.size() - 1;

        // move left while digits are 9
        while (i >= 0 && digits[i] == 9) {
            digits[i] = 0;   // make 9 → 0
            i--;
        }

        // if we stopped at a non-9
        if (i >= 0) {
            digits[i] += 1;
        } 
        else {
            // all digits were 9 → need extra digit
            digits.insert(digits.begin(), 1);
        }

        return digits;
    }
};

File: Python/66_Plus_One.py
from typing import List


class Solution:
    # Beat 100% in time and 47% in memory
    def plusOne(self, digits: List[int]) -> List[int]:

        index = len(digits) - 1

        while index >= 0 and digits[index] == 9:
            digits[index] = 0
            index = index - 1

        if index >= 0:
            digits[index] = digits[index] + 1
        else:
            # It shifts everything to the right so dont need to worry 
            digits.insert(0, 1)

        return digits